# 🚀 Alloy-Agent: Phi-3 Fine-Tuning for Steel Plant Maintenance

## Production-Grade Training Pipeline
- Model: Microsoft Phi-3 Mini (3.8B parameters)
- Method: QLoRA with Unsloth optimization
- Hardware: Google Colab T4 GPU (Free)
- Tracking: Weights & Biases

## Setup Instructions:
1. **Runtime**: Change runtime type → T4 GPU
2. **Execute**: Runtime → Run all (or run cells sequentially)
3. **Duration**: ~4-5 hours for full training

---

## 1. Install Dependencies
Installing Unsloth with compatible package versions for T4 GPU training.

In [1]:
# Install specific bitsandbytes version that works with Colab CUDA
!pip install -q -U "torchvision>=0.27.0"
!pip install -q -U "bitsandbytes>=0.48.0"  # Known working version for Colab
!pip install -q -U "unsloth[colab-new] @ git+https://github.com/unslothai/unsloth.git"
!pip install -q "transformers>=4.51.3,<=5.5.0" "datasets>=3.4.1,<4.4.0" "trl>=0.18.2,<=0.24.0"
!pip install -q -U accelerate peft wandb

# --- THE FIX ---
# 1. Remove conflicting CUDA 12 library left over by Colab's default environment
!pip uninstall -y nvidia-nvjitlink-cu12

# 2. Install the correct CUDA 13 library using NVIDIA's new package name
!pip install nvidia-nvjitlink

print("✅ Dependencies installed")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.6/7.6 MB 62.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 532.3/532.3 MB 3.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 366.2/366.2 MB 3.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 170.1/170.1 MB 6.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 206.0/206.0 MB 6.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.4/60.4 MB 14.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 201.5/201.5 MB 5.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.7/10.7 MB 119.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.2/90.2 MB 9.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 95.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 214.1/214.1 MB 6.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 79.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 59.5/59.

## 2. Fix CUDA Libraries
Setting up environment paths for bitsandbytes CUDA support.

In [2]:
import os
# Add specific CUDA 13.0 lib path for bitsandbytes
os.environ['LD_LIBRARY_PATH'] = f"/usr/local/cuda-13.0/lib64:{os.environ.get('LD_LIBRARY_PATH', '')}"
print("✅ CUDA environment configured")

✅ CUDA environment configured


## 3. Mount Google Drive
Connect to Drive and verify training data exists.

In [3]:
from google.colab import drive
import os

drive.mount('/content/drive')

TRAIN_PATH = '/content/drive/MyDrive/Alloy-Agent/data/training/train.jsonl'
VAL_PATH = '/content/drive/MyDrive/Alloy-Agent/data/training/val.jsonl'
OUTPUT_DIR = '/content/drive/MyDrive/Alloy-Agent/models/fine_tuned_phi3'

assert os.path.exists(TRAIN_PATH), f"Training data not found at {TRAIN_PATH}"
assert os.path.exists(VAL_PATH), f"Validation data not found at {VAL_PATH}"

print(f"✓ Data verified and ready")

Mounted at /content/drive
✓ Data verified and ready


## 3. Authenticate WandB
Login to Weights & Biases for experiment tracking.

In [4]:
import wandb

wandb.login()

/usr/local/lib/python3.12/dist-packages/notebook/notebookapp.py:191: SyntaxWarning: invalid escape sequence '\/'
  | |_| | '_ \/ _` / _` |  _/ -_)
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results


wandb: Enter your choice: 2


wandb: You chose 'Use an existing W&B account'
wandb: Logging into https://api.wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: Create a new API key at: https://wandb.ai/authorize?ref=models
wandb: Store your API key securely and do not share it.


wandb: Paste your API key and hit enter: ··········


wandb: No netrc file found, creating one.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: roxnazeer (abdul-nazeer) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


True

## 4. Load Model with LoRA
Loading Phi-3 Mini with 4-bit quantization and LoRA adapters for efficient fine-tuning.

In [5]:
from unsloth import FastLanguageModel
import torch

MAX_SEQ_LENGTH = 2048
DTYPE = None
LOAD_IN_4BIT = True

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name="unsloth/Phi-3-mini-4k-instruct",
    max_seq_length=MAX_SEQ_LENGTH,
    dtype=DTYPE,
    load_in_4bit=LOAD_IN_4BIT,
)

model = FastLanguageModel.get_peft_model(
    model,
    r=16,
    target_modules=["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj"],
    lora_alpha=16,
    lora_dropout=0.05,
    bias="none",
    use_gradient_checkpointing="unsloth",
    random_state=42,
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.


W0613 03:55:24.486000 597 torch/utils/_pytree.py:630] <enum 'KernelPreference'> is an Enum subclass and is now natively supported by torch.compile as an opaque value type. Calling register_constant() on Enum subclasses is deprecated and will be an error in a future release.
W0613 03:55:24.518000 597 torch/utils/_pytree.py:630] <enum 'ScaleCalculationMode'> is an Enum subclass and is now natively supported by torch.compile as an opaque value type. Calling register_constant() on Enum subclasses is deprecated and will be an error in a future release.


🦥 Unsloth Zoo will now patch everything to make training faster!


<string>:1: FutureWarning: torch._dynamo.config.inline_inbuilt_nn_modules is deprecated and does not do anything, inline_inbuilt_nn_modules is always True. It will be removed in a future version of PyTorch.


==((====))==  Unsloth 2026.6.6: Fast Mistral patching. Transformers: 5.5.0.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.563 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.12.0+cu130. CUDA: 7.5. CUDA Toolkit: 13.0. Triton: 3.7.0
\        /    Bfloat16 = FALSE. FA [Xformers = None. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


model.safetensors:   0%|          | 0.00/2.26G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/291 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/194 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.21k [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/3.34k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.84M [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/293 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/458 [00:00<?, ?B/s]

Unsloth: Dropout = 0 is supported for fast patching. You are using dropout = 0.05.
Unsloth will patch all other layers, except LoRA matrices, causing a performance hit.
Unsloth 2026.6.6 patched 32 layers with 0 QKV layers, 0 O layers and 0 MLP layers.


## 5. Load Datasets
Loading training and validation data from Google Drive.

In [6]:
from datasets import load_dataset

train_dataset = load_dataset('json', data_files=TRAIN_PATH, split='train')
val_dataset = load_dataset('json', data_files=VAL_PATH, split='train')

print(f"Train: {len(train_dataset):,} samples")
print(f"Val: {len(val_dataset):,} samples")
print(f"Columns: {train_dataset.column_names}")

Generating train split: 0 examples [00:00, ? examples/s]

Generating train split: 0 examples [00:00, ? examples/s]

Train: 1,776 samples
Val: 197 samples
Columns: ['instruction', 'input', 'output', 'metadata']


## 6. Tokenize Datasets
Preparing data for causal language modeling with Phi-3 chat format.

## 7. Training Configuration
Setting up hyperparameters: 3 epochs, batch size 2, gradient accumulation 4, learning rate 2e-4.

In [7]:
from trl import SFTConfig
import torch

# Define the configuration via SFTConfig to bypass the Pickling Bug
training_args = SFTConfig(
    output_dir="/content/drive/MyDrive/Alloy-Agent/checkpoints",
    num_train_epochs=3,
    per_device_train_batch_size=2,
    per_device_eval_batch_size=2,
    gradient_accumulation_steps=4,
    learning_rate=2e-4,
    weight_decay=0.01,
    warmup_steps=50,
    max_grad_norm=1.0,
    optim="adamw_torch",
    fp16=not torch.cuda.is_bf16_supported(),
    bf16=torch.cuda.is_bf16_supported(),
    logging_steps=10,
    logging_first_step=True,
    report_to="wandb",
    run_name="alloy-phi3-training",
    save_strategy="steps",
    save_steps=100,
    save_total_limit=3,
    eval_strategy="steps",
    eval_steps=100,
    load_best_model_at_end=True,
    metric_for_best_model="eval_loss",
    greater_is_better=False,
    gradient_checkpointing=True,
    gradient_checkpointing_kwargs={"use_reentrant": False},
    seed=42,
    data_seed=42,
    dataset_text_field="text",  # This line prevents the step 100 crash!
    max_seq_length=MAX_SEQ_LENGTH
)

print("✅ Section 7 updated with secure SFTConfig!")

✅ Section 7 updated with secure SFTConfig!


## 8. Initialize Trainer
Setting up trainer with early stopping (patience=3).

## 9. Start Training
Training will take approximately 4-5 hours on T4 GPU. Monitor progress at wandb.ai.

In [18]:
import torch
from trl import SFTTrainer, SFTConfig

# 1. Format your datasets correctly with the unified "text" field Unsloth expects
def formatting_prompts_func(examples):
    texts = []
    for instruction, input_text, output_text in zip(examples['instruction'], examples['input'], examples['output']):
        # Standardized ChatML / Industrial Agent Template
        text = (
            f"<|system|>You are an industrial maintenance AI assistant specialized in steel plant equipment analysis.<|end|>\n"
            f"<|user|>\n{instruction}\n{input_text}<|end|>\n"
            f"<|assistant|>\n{output_text}<|end|>"
        )
        texts.append(text)

    return { "text" : texts }

# Apply formatting mapping
train_dataset = train_dataset.map(formatting_prompts_func, batched=True)
val_dataset = val_dataset.map(formatting_prompts_func, batched=True)

# 2. Secure SFTConfig for Target Masking and Unsloth execution
sft_args = SFTConfig(
    output_dir="/content/drive/MyDrive/Alloy-Agent/checkpoints",
    num_train_epochs=3,
    per_device_train_batch_size=2,
    per_device_eval_batch_size=2,
    gradient_accumulation_steps=4,
    learning_rate=2e-4,
    weight_decay=0.01,
    warmup_steps=50,
    max_grad_norm=1.0,
    optim="adamw_torch",
    fp16=not torch.cuda.is_bf16_supported(),
    bf16=torch.cuda.is_bf16_supported(),
    logging_steps=10,
    logging_first_step=True,
    report_to="wandb",
    run_name="alloy-phi3-training",

    # Evaluation & Saving alignment (Critical for load_best_model_at_end)
    eval_strategy="steps",
    eval_steps=100,
    save_strategy="steps",
    save_steps=100,
    save_total_limit=3,
    load_best_model_at_end=True,
    metric_for_best_model="eval_loss",
    greater_is_better=False,

    gradient_checkpointing=True,
    gradient_checkpointing_kwargs={"use_reentrant": False},
    seed=42,
    data_seed=42,
    max_seq_length=MAX_SEQ_LENGTH,

    # --- UNSLOTH + TARGET MASKING CONFIGURATION ---
    dataset_text_field="text",
    completion_only_loss=True  # Enforces token masking on everything except assistant responses
)

# 3. Clean initialization of the SFTTrainer
trainer = SFTTrainer(
    model=model,
    tokenizer=tokenizer,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    args=sft_args
)

print("🚀 SFTTrainer perfectly configured and ready for production-grade training!")

Map:   0%|          | 0/1776 [00:00<?, ? examples/s]

Map:   0%|          | 0/197 [00:00<?, ? examples/s]

Unsloth: Tokenizing ["prompt"+"completion"] (num_proc=6):   0%|          | 0/1776 [00:00<?, ? examples/s]

Unsloth: Tokenizing ["prompt"+"completion"] (num_proc=6):   0%|          | 0/197 [00:00<?, ? examples/s]

🚀 SFTTrainer perfectly configured and ready for production-grade training!


In [19]:
import os
import wandb

# Force a clean start from Step 0 so the target masking takes full effect
checkpoint_dir = "/content/drive/MyDrive/Alloy-Agent/checkpoints"
print("Checking checkpoint directory...")

# Initialize tracking with the corrected configuration dictionary
wandb.init(
    project="alloy-agent-phi3",
    name="alloy-phi3-target-masked",
    config={
        "model": "Phi-3-mini-4k-instruct",
        "method": "QLoRA",
        "lora_r": 16,
        "batch_size": 8,
        "learning_rate": 2e-4,
        "epochs": 3,
        "enhancement": "DataCollatorForCompletionOnlyLM"
    }
)

print("🚀 Starting fresh optimized training run...")
# Enforce False to ensure we don't mix old unmasked weights
trainer.train(resume_from_checkpoint=False)

Checking checkpoint directory...


wandb: Detected [huggingface_hub.inference, openai] in use.
wandb: Use W&B Weave for improved LLM call tracing. Install Weave with `pip install weave` then add `import weave` to the top of your script.
wandb: For more information, check out the docs at: https://weave-docs.wandb.ai


🚀 Starting fresh optimized training run...


==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 1,776 | Num Epochs = 3 | Total steps = 666
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 29,884,416 of 3,850,963,968 (0.78% trained)
`use_return_dict` is deprecated! Use `return_dict` instead!


Unsloth: Will smartly offload gradients to save VRAM!
Unsloth: Double buffering enabled (parallel H2D + compute) for backward pass.


Step,Training Loss,Validation Loss
100,0.039712,0.031581
200,0.027243,0.026730
300,0.026298,0.026420
400,0.025635,0.025656
500,0.024583,0.025546
600,0.025025,0.025447
666,0.024622,0.025079


/usr/local/lib/python3.12/dist-packages/transformers/modeling_attn_mask_utils.py:71: FutureWarning: The attention mask API under `transformers.modeling_attn_mask_utils` (`AttentionMaskConverter`) is deprecated and will be removed in Transformers v5.10. Please use the new API in `transformers.masking_utils`.
  warnings.warn(DEPRECATION_MESSAGE, FutureWarning)
/usr/local/lib/python3.12/dist-packages/transformers/modeling_attn_mask_utils.py:281: FutureWarning: The attention mask API under `transformers.modeling_attn_mask_utils` (`AttentionMaskConverter`) is deprecated and will be removed in Transformers v5.10. Please use the new API in `transformers.masking_utils`.
  warnings.warn(DEPRECATION_MESSAGE, FutureWarning)
/usr/local/lib/python3.12/dist-packages/transformers/modeling_attn_mask_utils.py:254: FutureWarning: The attention mask API under `transformers.modeling_attn_mask_utils` (`AttentionMaskConverter`) is deprecated and will be removed in Transformers v5.10. Please use the new API 

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

Unsloth: Preserved sentencepiece asset `tokenizer.model` in /content/drive/MyDrive/Alloy-Agent/checkpoints/checkpoint-100.
/usr/local/lib/python3.12/dist-packages/transformers/modeling_attn_mask_utils.py:71: FutureWarning: The attention mask API under `transformers.modeling_attn_mask_utils` (`AttentionMaskConverter`) is deprecated and will be removed in Transformers v5.10. Please use the new API in `transformers.masking_utils`.
  warnings.warn(DEPRECATION_MESSAGE, FutureWarning)
/usr/local/lib/python3.12/dist-packages/transformers/modeling_attn_mask_utils.py:281: FutureWarning: The attention mask API under `transformers.modeling_attn_mask_utils` (`AttentionMaskConverter`) is deprecated and will be removed in Transformers v5.10. Please use the new API in `transformers.masking_utils`.
  warnings.warn(DEPRECATION_MESSAGE, FutureWarning)
/usr/local/lib/python3.12/dist-packages/transformers/modeling_attn_mask_utils.py:254: FutureWarning: The attention mask API under `transformers.modeling_a

TrainOutput(global_step=666, training_loss=0.09429083265610286, metrics={'train_runtime': 7133.7322, 'train_samples_per_second': 0.747, 'train_steps_per_second': 0.093, 'total_flos': 8.565113195071488e+16, 'train_loss': 0.09429083265610286, 'epoch': 3.0})

In [20]:
# 1. Save the optimized LoRA adapters and tokenizer
model.save_pretrained_merged(
    "/content/drive/MyDrive/Alloy-Agent/models/production_phi3_masked",
    tokenizer,
    save_method="lora"
)

print("💾 Production-grade target-masked weights successfully locked into Google Drive!")

config.json:   0%|          | 0.00/724 [00:00<?, ?B/s]

Unsloth: Restored added_tokens_decoder metadata in /content/drive/MyDrive/Alloy-Agent/models/production_phi3_masked/tokenizer_config.json.
Unsloth: Preserved sentencepiece asset `tokenizer.model` in /content/drive/MyDrive/Alloy-Agent/models/production_phi3_masked.


Found HuggingFace hub cache directory: /root/.cache/huggingface/hub


Fetching 1 files:   0%|          | 0/1 [00:00<?, ?it/s]

Checking cache directory for required files...
Cache check failed: model-00001-of-00002.safetensors not found in local cache.
Not all required files found in cache. Will proceed with downloading.
Checking cache directory for required files...
Cache check failed: tokenizer.model not found in local cache.
Not all required files found in cache. Will proceed with downloading.



Unsloth: Preparing safetensor model files:   0%|          | 0/2 [00:00<?, ?it/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/4.99G [00:00<?, ?B/s]


Unsloth: Preparing safetensor model files:  50%|█████     | 1/2 [02:21<02:21, 141.02s/it]

model-00002-of-00002.safetensors:   0%|          | 0.00/2.65G [00:00<?, ?B/s]


Unsloth: Preparing safetensor model files: 100%|██████████| 2/2 [03:26<00:00, 103.20s/it]

Unsloth: Merging weights into 16bit: 100%|██████████| 2/2 [05:33<00:00, 166.81s/it]


Unsloth: Merge process complete. Saved to `/content/drive/MyDrive/Alloy-Agent/models/production_phi3_masked`
💾 Production-grade target-masked weights successfully locked into Google Drive!


## 10. Save Model
Saving fine-tuned LoRA adapters to Google Drive.

In [21]:
!pip install -q huggingface_hub

In [22]:
import os
from huggingface_hub import HfApi, login

# 1. Authenticate securely using your token
HF_TOKEN = "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"
print("🔐 Logging into Hugging Face Hub...")
login(token=HF_TOKEN)

# 2. Define configuration with your credentials
HF_USERNAME = "CodeMasterAbdul"
REPO_NAME = "alloy-phi3-steel-maintenance"
LOCAL_DIR = "/content/drive/MyDrive/Alloy-Agent/models/production_phi3_masked"
repo_id = f"{HF_USERNAME}/{REPO_NAME}"

# 3. Create the repository on your profile (if it doesn't exist yet)
print(f"🔄 Creating repository {repo_id}...")
api = HfApi()
api.create_repo(repo_id=repo_id, private=False, exist_ok=True)

# 4. Upload the production folder directly from Google Drive
print(f"📤 Uploading merged 16-bit model weights to Hugging Face...")
api.upload_folder(
    folder_path=LOCAL_DIR,
    repo_id=repo_id,
    repo_type="model"
)

print(f"🎉 Success! Your production model is live at: https://huggingface.co/{repo_id}")

🔐 Logging into Hugging Face Hub...
🔄 Creating repository CodeMasterAbdul/alloy-phi3-steel-maintenance...
📤 Uploading merged 16-bit model weights to Hugging Face...


Processing Files (0 / 0)      : |          |  0.00B /  0.00B            

New Data Upload               : |          |  0.00B /  0.00B            

  ...i3_masked/tokenizer.model: 100%|##########|  500kB /  500kB            

  ...0002-of-00002.safetensors:   1%|1         | 32.0MB / 2.65GB            

  ...0001-of-00002.safetensors:   1%|          | 32.0MB / 4.99GB            

🎉 Success! Your production model is live at: https://huggingface.co/CodeMasterAbdul/alloy-phi3-steel-maintenance


## 11. Test Inference
Quick test to verify the fine-tuned model works correctly.

In [23]:
from unsloth import FastLanguageModel

# 1. Activate ultra-fast native inference mode
FastLanguageModel.for_inference(model)

# 2. Structure a test prompt matching your industrial template exactly
messages = [
    {
        "role": "system",
        "content": "You are an industrial maintenance AI assistant specialized in steel plant equipment analysis."
    },
    {
        "role": "user",
        "content": "The blast furnace cooling plate is showing a sudden temperature spike. What are the immediate troubleshooting steps?"
    }
]

inputs = tokenizer.apply_chat_template(
    messages,
    tokenize=True,
    add_generation_prompt=True,
    return_tensors="pt"
).to("cuda")

# 3. Generate the response
print("🤖 Generating maintenance analysis...")
outputs = model.generate(
    input_ids=inputs,
    max_new_tokens=256,
    use_cache=True,
    temperature=0.1, # Keep it deterministic and precise
    top_p=0.9
)

# 4. Extract and print only the model's response
generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
print("\n" + generated_text.split("<|assistant|>")[-1].strip())

The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.


🤖 Generating maintenance analysis...


Both `max_new_tokens` (=256) and `max_length`(=4096) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
/usr/local/lib/python3.12/dist-packages/transformers/modeling_attn_mask_utils.py:71: FutureWarning: The attention mask API under `transformers.modeling_attn_mask_utils` (`AttentionMaskConverter`) is deprecated and will be removed in Transformers v5.10. Please use the new API in `transformers.masking_utils`.
  warnings.warn(DEPRECATION_MESSAGE, FutureWarning)
/usr/local/lib/python3.12/dist-packages/transformers/modeling_attn_mask_utils.py:281: FutureWarning: The attention mask API under `transformers.modeling_attn_mask_utils` (`AttentionMaskConverter`) is deprecated and will be removed in Transformers v5.10. Please use the new API in `transformers.masking_utils`.
  warnings.warn(DEPRECATION_MESSAGE, FutureWarning)
/usr/local/lib/python3.12/


You are an industrial maintenance AI assistant specialized in steel plant equipment analysis. The blast furnace cooling plate is showing a sudden temperature spike. What are the immediate troubleshooting steps? 1. Immediately reduce the cooling water flow to minimize thermal stress.

2. Increase monitoring frequency to capture temperature trends.

3. Check for blockages in the cooling system.

4. Inspect the blast furnace for insulation degradation or heat leakage.

5. Evaluate the possibility of chemical reaction exothermicity.

6. Review recent maintenance activities for causative actions.

7. Prepare to perform a detailed thermodynamic analysis.


In [24]:
import wandb
wandb.finish()

eval/loss,█▃▂▂▂▁▁
eval/runtime,▆█▇▁█▁▄
eval/samples_per_second,▃▁▂█▁█▆
eval/steps_per_second,▃▁▂▇▂█▅
train/epoch,▁▁▁▂▂▂▂▃▃▃▃▃▃▄▄▄▄▄▄▅▅▅▆▆▆▆▆▆▆▇▇▇▇▇▇█████
train/global_step,▁▁▁▁▂▂▂▂▂▂▂▃▃▃▃▃▃▄▄▄▄▄▅▅▅▅▆▆▆▆▆▇▇▇▇▇████
train/grad_norm,█▇▅▅▄▃▄▃▂▃▂▁▂▂▂▂▂▂▂▁▂▁▁▁▁▁▁▁▁▁▁▂▁▁▁▁▁▁▁▁
train/learning_rate,▁▂▄▅▇███▇▇▇▇▆▆▆▆▆▆▅▅▅▅▅▄▄▄▄▄▃▃▃▃▃▃▂▂▂▂▁▁
train/loss,██▇▅▃▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁
eval/loss,0.02508
eval/runtime,80.7516


## Training Summary

**Next Steps:**
1. Push model to HuggingFace Hub for deployment
2. Use HF Inference API endpoint (free hosting)
3. Integrate with Alloy-Agent backend

**Model Details:**
- Base: Phi-3-mini-4k-instruct (3.8B parameters)
- Method: QLoRA with 4-bit quantization
- LoRA rank: 16, alpha: 16
- Training: 3 epochs, ~1,776 samples
- Saved: Google Drive at `/Alloy-Agent/models/fine_tuned_phi3`